### ISSUE TITLE
Unable to process module devicemonitor add or update as the module identity could not be obtained"

### ISSUE THREAD

**Issue by Rollin-stone** [Azure/iotedge#5207]
Hi,

I'm facing a problem since 2 weeks.

Sometime when I want to deploy a new version of a module on the IoT Edge device I have this message in the IoT EdgeAgent logs

"Unable to process module devicemonitor add or update as the module identity could not be obtained" .

Sometimes I have the same problem with the others modules (5 modules ), sometimes all at the same time.

the systemctl restart ioetdge on the device fixes the pb but when I want to deploy a new version, same error.

After restart the edgeAgent logs

Scraping frequency: 01:00:00
Upload Frequency: 1.00:00:00
<6> 2021-06-28 15:56:34.039 +00:00 [INF] - Registering request handler UploadModuleLogs
<6> 2021-06-28 15:56:34.040 +00:00 [INF] - Registering request handler GetModuleLogs
<6> 2021-06-28 15:56:34.041 +00:00 [INF] - Registering request handler UploadSupportBundle
<6> 2021-06-28 15:56:34.042 +00:00 [INF] - Registering request handler RestartModule
<4> 2021-06-28 15:56:40.074 +00:00 [WRN] - Empty edge agent config was received. Attempting to read config from backup (/tmp/edgeAgent/backup.json) instead
<6> 2021-06-28 15:56:40.179 +00:00 [INF] - Obtained edge agent config from backup config file - /tmp/edgeAgent/backup.json
<6> 2021-06-28 15:56:41.575 +00:00 [INF] - Edge agent connected to IoT Hub via Amqp_Tcp_Only.
<6> 2021-06-28 15:56:41.954 +00:00 [INF] - Initialized new module client with subscriptions enabled
<6> 2021-06-28 15:56:42.120 +00:00 [INF] - Obtained Edge agent twin from IoTHub with desired properties version 42 and reported properties version 358.
<6> 2021-06-28 15:56:49.622 +00:00 [INF] - Plan execution started for deployment 42
<6> 2021-06-28 15:56:49.637 +00:00 [INF] - Executing command: "Saving devicemonitor to store"
<6> 2021-06-28 15:56:49.813 +00:00 [INF] - Executing command: "Command Group: (\n [Command Group: (\n [Prepare update module devicemonitor]\n [Stop module devicemonitor]\n [Update module devicemonitor]\n)]\n [Start module devicemonitor]\n)"
<6> 2021-06-28 15:56:49.821 +00:00 [INF] - Executing command: "Command Group: (\n [Prepare update module devicemonitor]\n [Stop module devicemonitor]\n [Update module devicemonitor]\n)"
<6> 2021-06-28 15:59:13.549 +00:00 [INF] - Executing command: "Start module devicemonitor"
<6> 2021-06-28 15:59:14.907 +00:00 [INF] - Executing command: "Command Group: (\n [Stop module edgeHub]\n [Start module edgeHub]\n [Saving edgeHub to store]\n

My devices are based on raspberry os and connected to IoT Central

Any idea ?

uname -a 
--------------------------
Linux xxxx 4.9.75-axo-r-KC01+ #3 SMP Tue Jan 30 12:12:26 CET 2018 armv7l GNU/Linux

iotedge version 
------------------
iotedge 1.1.3

Configuration checks
--------------------
√ config.yaml is well-formed - OK
√ config.yaml has well-formed connection string - OK
√ container engine is installed and functional - OK
√ config.yaml has correct hostname - OK
√ config.yaml has correct URIs for daemon mgmt endpoint - OK
√ latest security daemon - OK
‼ host time is close to real time - Warning
    Time on the device is out of sync with the NTP server. This may cause problems connecting to IoT Hub.
    Please ensure time on device is accurate, for example by installing an NTP daemon.
√ container time is close to host time - OK
‼ DNS server - Warning
    Container engine is not configured with DNS server setting, which may impact connectivity to IoT Hub.
    Please see https://aka.ms/iotedge-prod-checklist-dns for best practices.
    You can ignore this warning if you are setting DNS server per module in the Edge deployment.
‼ production readiness: certificates - Warning
    The Edge device is using self-signed automatically-generated development certificates.
    They will expire in 82 days (at 2021-09-26 10:09:49 UTC) causing module-to-module and downstream device communication to fail on an active deployment.
    After the certs have expired, restarting the IoT Edge daemon will trigger it to generate new development certs.
    Please consider using production certificates instead. See https://aka.ms/iotedge-prod-checklist-certs for best practices.
√ production readiness: container engine - OK
‼ production readiness: logs policy - Warning
    Container engine is not configured to rotate module logs which may cause it run out of disk space.
    Please see https://aka.ms/iotedge-prod-checklist-logs for best practices.
    You can ignore this warning if you are setting log policy per module in the Edge deployment.
‼ production readiness: Edge Agent's storage directory is persisted on the host filesystem - Warning
    The edgeAgent module is not configured to persist its /tmp/edgeAgent directory on the host filesystem.
    Data might be lost if the module is deleted or updated.
    Please see https://aka.ms/iotedge-storage-host for best practices.
‼ production readiness: Edge Hub's storage directory is persisted on the host filesystem - Warning
    The edgeHub module is not configured to persist its /tmp/edgeHub directory on the host filesystem.
    Data might be lost if the module is deleted or updated.
    Please see https://aka.ms/iotedge-storage-host for best practices.

Connectivity checks
-------------------
√ host can connect to and perform TLS handshake with DPS endpoint - OK
√ host can connect to and perform TLS handshake with IoT Hub AMQP port - OK
√ host can connect to and perform TLS handshake with IoT Hub HTTPS / WebSockets port - OK
√ host can connect to and perform TLS handshake with IoT Hub MQTT port - OK
√ container on the default network can connect to IoT Hub AMQP port - OK
√ container on the default network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the default network can connect to IoT Hub MQTT port - OK
√ container on the IoT Edge module network can connect to IoT Hub AMQP port - OK
√ container on the IoT Edge module network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the IoT Edge module network can connect to IoT Hub MQTT port - OK

18 check(s) succeeded.
6 check(s) raised warnings. Re-run with --verbose for more details.

Thanks;

**Comment by asergaz** [comment]
This is the related thread initiated in Microsoft Q&A: https://docs.microsoft.com/en-us/answers/questions/454868/index.html

@Rollin-stone note the warning "host time is close to real time" can you confirm if you can fix it and try to reproduce the issue? I suspect that may be the cause of the issue.

**Comment by vadim-kovalyov** [comment]
@Rollin-stone, have you tried what @asergaz has suggested above?

**Comment by Rollin-stone** [comment]
Hi @asergaz,

I didn't understand what you meant. Is it about the time sync ? 
yes I I synchronised the device (base on raspberry) with the ntp server but I have the same pb.

I have this error (in edgeAgent logs)  when I push a new deployment manifest to the device.

I have updated de deployment manifest with a new version of the "devicemonitor" module. Only way to fix it is to restart iotedge (systemctl restart iotedge) until the next update...

[support_bundle.zip](https://github.com/Azure/iotedge/files/6791018/support_bundle.zip)

2021-07-09 09:36:17.711 +00:00 [INF] - Successfully completed periodic operation Metrics Scrape
<6> 2021-07-09 10:36:13.731 +00:00 [INF] - Starting periodic operation refresh twin config...
<6> 2021-07-09 10:36:13.767 +00:00 [INF] - Obtained Edge agent twin from IoTHub with desired properties version 49 and reported properties version 471.
<6> 2021-07-09 10:36:13.774 +00:00 [INF] - Successfully completed periodic operation refresh twin config
<6> 2021-07-09 10:36:17.708 +00:00 [INF] - Starting periodic operation Metrics Scrape...
<6> 2021-07-09 10:36:17.708 +00:00 [INF] - Scraping Metrics
<6> 2021-07-09 10:36:17.708 +00:00 [INF] - Scraping endpoint http://edgeHub:9600/metrics
<6> 2021-07-09 10:36:17.713 +00:00 [INF] - Scraping endpoint http://edgeAgent:9600/metrics
<6> 2021-07-09 10:36:17.800 +00:00 [INF] - Storing Metrics
<6> 2021-07-09 10:36:17.883 +00:00 [INF] - Scraped and Stored Metrics
<6> 2021-07-09 10:36:17.883 +00:00 [INF] - Successfully completed periodic operation Metrics Scrape
<6> 2021-07-09 10:46:09.758 +00:00 [INF] - Obtained Edge agent twin from IoTHub with desired properties version 49 and reported properties version 471.
<6> 2021-07-09 10:52:24.107 +00:00 [INF] - Unable to process module devicemonitor add or update as the module identity could not be obtained
<6> 2021-07-09 10:52:24.112 +00:00 [INF] - Plan execution started for deployment 50
<6> 2021-07-09 10:52:24.112 +00:00 [INF] - Executing command: "Saving devicemonitor to store"
<6> 2021-07-09 10:52:24.113 +00:00 [INF] - Plan execution ended for deployment 50
<6> 2021-07-09 10:52:24.358 +00:00 [INF] - Updated reported properties
<6> 2021-07-09 10:52:45.598 +00:00 [INF] - Unable to process module devicemonitor add or update as the module identity could not be obtained
<6> 2021-07-09 10:52:45.601 +00:00 [INF] - Plan execution started for deployment 50
<6> 2021-07-09 10:52:45.602 +00:00 [INF] - Executing command: "Saving devicemonitor to store"
<6> 2021-07-09 10:52:45.604 +00:00 [INF] - Plan execution ended for deployment 50
<6> 2021-

**Comment by asergaz** [comment]
@vadim-kovalyov do you have further suggestions for troubleshoot? Thanks

**Comment by vadim-kovalyov** [comment]
I did not find anything specific in the support bundle, and was not able to repro it myself. So I'm still looking into the source code, but so far no luck. Two thing that we should try:
- @Rollin-stone, I see that you use 1.0 version of edgeAgent, can you upgrade to 1.1 (in the deployment manifest) and see if it helps?
-  If you can still repro the issue, can you please enable debug level logs in edgeAgent (by setting env var `RuntimeLogLevel=debug` and collect another support bundle?

**Comment by vadim-kovalyov** [comment]
I'll close this for now due to inactivity. Please re-open again and @ me if issue still exists after updating to 1.1 and when you have the support bundle with debug logs.

**Comment by milijan** [comment]
@vadim-kovalyov it seems I ran in a similar issue on Eflow. Couldn't reproduce more than once as I have limited hardware to try it on. But the steps go something like this:
- Deploy eflow with WAC
- manually change edgeAgent/EdgeHub version to 1.2 on the Azure Portal Runtime settings.
- realising the mistake (Eflow only supports 1.1LTS) and putting the version back to 1.1
- uninstalling Eflow/reprovisioning/restarting has no effect.
- now stuck

> sudo iotedge restart edgeHub
A module runtime error occurred
        caused by: Could not restart module edgeHub
        caused by: Client error

no matter what I do edgeHub is not coming back up. Any idea?
